core/editor.js

```diff
@@ -168,8 +168,23 @@
 		} );
 		this.on( 'mode', updateCommands );
 
+		// (#3175)
+		this.on( 'selectionCheck', function( evt ) {
+			if ( evt.data ) {
+				evt.data.optimizeInElementEnds();
+			}
+		} );
+
 		// Handle startup focus.
 		this.on( 'instanceReady', function() {
+			this.editable().on( 'keydown', function( evt ) {
+				this._.shiftPressed = evt.data.$.shiftKey;
+			}, this );
+
+			this.editable().on( 'keyup', function( evt ) {
+				this._.shiftPressed = evt.data.$.shiftKey;
+			}, this );
+
```

Comment:
I would be for moving all the changes from this file to `core/selection/optimization.js `, and here just add something like:

```js
// (#3175)
this.on( 'instanceReady', function( evt ) {
    CKEDITOR.dom.selection.setupOptimizations( evt.editor );
} );
```

and `setupOptimizations();` would attach all listeners. WDYT?